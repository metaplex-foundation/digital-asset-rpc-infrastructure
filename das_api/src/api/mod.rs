use crate::DasApiError;
use async_trait::async_trait;
use digital_asset_types::rpc::filter::SearchConditionType;
use digital_asset_types::rpc::response::AssetList;
use digital_asset_types::rpc::{filter::AssetSorting, response::GetGroupingResponse};
use digital_asset_types::rpc::{Asset, AssetProof, Interface, OwnershipModel, RoyaltyModel};
use open_rpc_derive::{document_rpc, rpc};
use open_rpc_schema::schemars::JsonSchema;
use serde::{Deserialize, Serialize};

mod api_impl;
pub use api_impl::*;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAssetsByGroup {
    pub group_key: String,
    pub group_value: String,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAssetsByOwner {
    pub owner_address: String,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAsset {
    pub id: String,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]

pub struct GetAssetsByCreator {
    pub creator_address: String,
    pub only_verified: Option<bool>,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct SearchAssets {
    pub negate: Option<bool>,
    pub condition_type: Option<SearchConditionType>,
    pub interface: Option<Interface>,
    pub owner_address: Option<String>,
    pub owner_type: Option<OwnershipModel>,
    pub creator_address: Option<String>,
    pub creator_verified: Option<bool>,
    pub authority_address: Option<String>,
    pub grouping: Option<(String, String)>,
    pub delegate: Option<Vec<u8>>,
    pub frozen: Option<bool>,
    pub supply: Option<u64>,
    pub supply_mint: Option<String>,
    pub compressed: Option<bool>,
    pub compressible: Option<bool>,
    pub royalty_target_type: Option<RoyaltyModel>,
    pub royalty_target: Option<String>,
    pub royalty_amount: Option<u32>,
    pub burnt: Option<bool>,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
    #[serde(default)]
    pub json_uri: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAssetsByAuthority {
    pub authority_address: String,
    pub sort_by: Option<AssetSorting>,
    pub limit: Option<u32>,
    pub page: Option<u32>,
    pub before: Option<String>,
    pub after: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetGrouping {
    pub group_key: String,
    pub group_value: String,
}

#[document_rpc]
#[async_trait]
pub trait ApiContract: Send + Sync + 'static {
    async fn check_health(&self) -> Result<(), DasApiError>;
    #[rpc(
        name = "getAssetProof",
        params = "named",
        summary = "Get a merkle proof for a compressed asset by its ID"
    )]
    async fn get_asset_proof(&self, payload: GetAsset) -> Result<AssetProof, DasApiError>;
    #[rpc(
        name = "getAsset",
        params = "named",
        summary = "Get an asset by its ID"
    )]
    async fn get_asset(&self, payload: GetAsset) -> Result<Asset, DasApiError>;
    #[rpc(
        name = "getAssetsByOwner",
        params = "named",
        summary = "Get a list of assets owned by an address"
    )]
    async fn get_assets_by_owner(
        &self,
        payload: GetAssetsByOwner,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getAssetsByGroup",
        params = "named",
        summary = "Get a list of assets by a group key and value"
    )]
    async fn get_assets_by_group(
        &self,
        payload: GetAssetsByGroup,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getAssetsByCreator",
        params = "named",
        summary = "Get a list of assets created by an address"
    )]
    async fn get_assets_by_creator(
        &self,
        payload: GetAssetsByCreator,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getAssetsByAuthority",
        params = "named",
        summary = "Get a list of assets with a specific authority"
    )]
    async fn get_assets_by_authority(
        &self,
        payload: GetAssetsByAuthority,
    ) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "searchAssets",
        params = "named",
        summary = "Search for assets by a variety of parameters"
    )]
    async fn search_assets(&self, payload: SearchAssets) -> Result<AssetList, DasApiError>;
    #[rpc(
        name = "getGrouping",
        params = "named",
        summary = "Get a list of assets grouped by a specific authority"
    )]
    async fn get_grouping(&self, payload: GetGrouping) -> Result<GetGroupingResponse, DasApiError>;
}
