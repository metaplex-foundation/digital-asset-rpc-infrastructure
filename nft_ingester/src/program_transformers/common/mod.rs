use crate::IngesterError;
use blockbuster::programs::bubblegum::ChangeLogEventV1;
use digital_asset_types::dao::{backfill_items, cl_items};
use sea_orm::{entity::*, query::*, sea_query::OnConflict, DatabaseTransaction, DbBackend};

pub mod task;

pub async fn save_changelog_event(
    change_log_event: &ChangeLogEventV1,
    slot: u64,
    txn: &DatabaseTransaction,
) -> Result<u64, IngesterError> {
    insert_change_log(change_log_event, slot, txn, false).await?;
    Ok(change_log_event.seq)
}

fn node_idx_to_leaf_idx(index: i64, tree_height: u32) -> i64 {
    index - 2i64.pow(tree_height)
}

pub async fn insert_change_log(
    change_log_event: &ChangeLogEventV1,
    slot: u64,
    txn: &DatabaseTransaction,
    filling: bool,
) -> Result<(), IngesterError> {
    let mut i: i64 = 0;
    let depth = change_log_event.path.len() - 1;
    let tree_id = change_log_event.id.as_ref();
    for p in change_log_event.path.iter() {
        let node_idx = p.index as i64;
        println!(
            "seq {}, index {} level {}, node {:?}",
            change_log_event.seq,
            p.index,
            i,
            bs58::encode(p.node).into_string()
        );
        let leaf_idx = if i == 0 {
            Some(node_idx_to_leaf_idx(node_idx, depth as u32))
        } else {
            None
        };

        let item = cl_items::ActiveModel {
            tree: Set(tree_id.to_vec()),
            level: Set(i),
            node_idx: Set(node_idx),
            hash: Set(p.node.as_ref().to_vec()),
            seq: Set(change_log_event.seq as i64),
            leaf_idx: Set(leaf_idx),
            ..Default::default()
        };
        i += 1;
        let mut query = cl_items::Entity::insert(item)
            .on_conflict(
                OnConflict::columns([cl_items::Column::Tree, cl_items::Column::NodeIdx])
                    .update_columns([cl_items::Column::Hash, cl_items::Column::Seq])
                    .to_owned(),
            )
            .build(DbBackend::Postgres);
        if !filling {
            query.sql = format!("{} WHERE excluded.seq > cl_items.seq", query.sql);
        }
        txn.execute(query)
            .await
            .map_err(|db_err| IngesterError::StorageWriteError(db_err.to_string()))?;
    }

    // If and only if the entire path of nodes was inserted into the `cl_items` table, then insert
    // a single row into the `backfill_items` table.  This way if an incomplete path was inserted
    // into `cl_items` due to an error, a gap will be created for the tree and the backfiller will
    // fix it.
    if i - 1 == depth as i64 {
        // See if the tree already exists in the `backfill_items` table.
        let rows = backfill_items::Entity::find()
            .filter(backfill_items::Column::Tree.eq(tree_id))
            .limit(1)
            .all(txn)
            .await?;

        // If the tree does not exist in `backfill_items` and the sequence number is greater than 1,
        // then we know we will need to backfill the tree from sequence number 1 up to the current
        // sequence number.  So in this case we set at flag to force checking the tree.
        let force_chk = rows.is_empty() && change_log_event.seq > 1;

        println!("Adding to backfill_items table at level {}", i - 1);
        let item = backfill_items::ActiveModel {
            tree: Set(tree_id.to_vec()),
            seq: Set(change_log_event.seq as i64),
            slot: Set(slot as i64),
            force_chk: Set(force_chk),
            backfilled: Set(false),
            failed: Set(false),
            ..Default::default()
        };

        backfill_items::Entity::insert(item).exec(txn).await?;
    }

    Ok(())
    //TODO -> set maximum size of path and break into multiple statements
}
