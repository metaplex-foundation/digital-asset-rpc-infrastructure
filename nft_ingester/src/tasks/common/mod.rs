use {
    super::{BgTask, FromTaskData, IngesterError, IntoTaskData, TaskData},
    async_trait::async_trait,
    chrono::{NaiveDateTime, Utc},
    digital_asset_types::dao::asset_data,
    futures::future::BoxFuture,
    log::debug,
    program_transformers::{DownloadMetadataInfo, DownloadMetadataNotifier},
    reqwest::{Client, ClientBuilder},
    sea_orm::*,
    serde::{Deserialize, Serialize},
    std::{
        fmt::{Display, Formatter},
        time::Duration,
    },
    tokio::sync::mpsc::UnboundedSender,
    url::Url,
};

pub fn create_download_metadata_notifier(
    bg_task_sender: UnboundedSender<TaskData>,
) -> DownloadMetadataNotifier {
    Box::new(
        move |info: DownloadMetadataInfo| -> BoxFuture<
            'static,
            Result<(), Box<dyn std::error::Error + Send + Sync>>,
        > {
            let (asset_data_id, uri) = info.into_inner();
            let task = DownloadMetadata {
                asset_data_id,
                uri,
                created_at: Some(Utc::now().naive_utc()),
            };
            let task = task
                .into_task_data()
                .and_then(|task| {
                    bg_task_sender.send(task).map_err(Into::into)
                })
                .map_err(Into::into);
            Box::pin(async move { task })
        },
    )
}

const TASK_NAME: &str = "DownloadMetadata";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadMetadata {
    pub asset_data_id: Vec<u8>,
    pub uri: String,
    #[serde(skip_serializing)]
    pub created_at: Option<NaiveDateTime>,
}

impl DownloadMetadata {
    pub fn sanitize(&mut self) {
        self.uri = self.uri.trim().replace('\0', "");
    }
}

impl IntoTaskData for DownloadMetadata {
    fn into_task_data(self) -> Result<TaskData, IngesterError> {
        let ts = self.created_at;
        let data =
            serde_json::to_value(self).map_err(<serde_json::Error as Into<IngesterError>>::into)?;
        Ok(TaskData {
            name: TASK_NAME,
            data,
            created_at: ts,
        })
    }
}

impl FromTaskData<DownloadMetadata> for DownloadMetadata {
    fn from_task_data(data: TaskData) -> Result<Self, IngesterError> {
        serde_json::from_value(data.data).map_err(|e| e.into())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadMetadataTask {
    pub lock_duration: Option<i64>,
    pub max_attempts: Option<i16>,
    pub timeout: Option<Duration>,
}

impl DownloadMetadataTask {
    async fn request_metadata(
        uri: String,
        timeout: Duration,
    ) -> Result<serde_json::Value, IngesterError> {
        let client = ClientBuilder::new().timeout(timeout).build()?;
        let response = Client::get(&client, uri) // Need to check for malicious sites ?
            .send()
            .await?;

        if response.status() != reqwest::StatusCode::OK {
            Err(IngesterError::HttpError {
                status_code: response.status().as_str().to_string(),
            })
        } else {
            let val: serde_json::Value = response.json().await?;
            Ok(val)
        }
    }
}

#[derive(FromQueryResult, Debug, Default, Clone, Eq, PartialEq)]
struct MetadataUrl {
    pub metadata_url: String,
}

#[async_trait]
impl BgTask for DownloadMetadataTask {
    fn name(&self) -> &'static str {
        TASK_NAME
    }

    fn lock_duration(&self) -> i64 {
        self.lock_duration.unwrap_or(5)
    }

    fn max_attempts(&self) -> i16 {
        self.max_attempts.unwrap_or(3)
    }

    async fn task(
        &self,
        db: &DatabaseConnection,
        data: serde_json::Value,
    ) -> Result<(), IngesterError> {
        let download_metadata: DownloadMetadata = serde_json::from_value(data)?;
        let meta_url = Url::parse(&download_metadata.uri);
        let body = match meta_url {
            Ok(_) => {
                DownloadMetadataTask::request_metadata(
                    download_metadata.uri.clone(),
                    self.timeout.unwrap_or(Duration::from_secs(3)),
                )
                .await?
            }
            _ => serde_json::Value::String("Invalid Uri".to_string()), //TODO -> enumize this.
        };

        let query = asset_data::Entity::find_by_id(download_metadata.asset_data_id.clone())
            .select_only()
            .column(asset_data::Column::MetadataUrl)
            .build(DbBackend::Postgres);

        match MetadataUrl::find_by_statement(query).one(db).await? {
            Some(asset) => {
                if asset.metadata_url != download_metadata.uri {
                    debug!(
                        "skipping download metadata of old URI for {:?}",
                        bs58::encode(download_metadata.asset_data_id.clone()).into_string()
                    );
                    return Ok(());
                }
            }
            None => {
                return Err(IngesterError::UnrecoverableTaskError(format!(
                    "failed to find URI in database for {:?}",
                    bs58::encode(download_metadata.asset_data_id.clone()).into_string()
                )));
            }
        }

        let model = asset_data::ActiveModel {
            id: Unchanged(download_metadata.asset_data_id.clone()),
            metadata: Set(body),
            reindex: Set(Some(false)),
            ..Default::default()
        };
        debug!(
            "download metadata for {:?}",
            bs58::encode(download_metadata.asset_data_id.clone()).into_string()
        );
        asset_data::Entity::update(model)
            .filter(asset_data::Column::Id.eq(download_metadata.asset_data_id.clone()))
            .filter(
                Condition::all()
                    .add(asset_data::Column::MetadataUrl.eq(download_metadata.uri.clone())),
            )
            .exec(db)
            .await
            .map(|_| ())
            .map_err(|db| {
                IngesterError::TaskManagerError(format!(
                    "Database error with {}, error: {}",
                    self.name(),
                    db
                ))
            })?;

        if meta_url.is_err() {
            return Err(IngesterError::UnrecoverableTaskError(format!(
                "Failed to parse URI: {}",
                download_metadata.uri
            )));
        }
        Ok(())
    }
}

impl Display for DownloadMetadata {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "DownloadMetadata from {} for {:?}",
            self.uri, self.asset_data_id
        )
    }
}
