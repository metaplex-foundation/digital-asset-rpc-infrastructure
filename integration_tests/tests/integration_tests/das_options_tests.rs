use function_name::named;

use das_api::api::{self, ApiContract};

use itertools::Itertools;

use serial_test::serial;

use super::common::*;

#[tokio::test]
#[serial]
#[named]

async fn test_get_asset_with_show_collection_metadata_option() {
    let name = trim_test_name(function_name!());

    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "AH6wj7T8Ke5nbukjtcobjjs1CDWUcQxndtnLkKAdrSrM",
        "3crBqZZsHhoLphM55MG4KRW6SbNzFEBFnehw7PW7ZRKt",
        "7fXKY9tPpvYsdbSNyesUqo27WYC6ZsBEULdtngGHqLCK",
        "J1S9H3QjnRtBbbuD4HjPV6RpRhwuk4zKbxsnCHuTgh9w",
        "8KyuwGzav7jTW9YaBGj2Qtp2q24zPUR3rD5caojXaby4",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;

    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    index_metadata_jsons(
        &setup,
        &[
            "AH6wj7T8Ke5nbukjtcobjjs1CDWUcQxndtnLkKAdrSrM",
            "J1S9H3QjnRtBbbuD4HjPV6RpRhwuk4zKbxsnCHuTgh9w",
        ],
        serde_json::from_str(SAMPLE_METADATA_JSON).unwrap(),
    )
    .await;

    let request = r#"
    {
    "id": "AH6wj7T8Ke5nbukjtcobjjs1CDWUcQxndtnLkKAdrSrM",
    "displayOptions" : {
        "showCollectionMetadata": true
        }
    }
    "#;

    let request: api::GetAsset = serde_json::from_str(request).unwrap();

    let response = setup.das_api.get_asset(request).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_get_asset_by_group_with_show_collection_metadata_option() {
    let name = trim_test_name(function_name!());

    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "AH6wj7T8Ke5nbukjtcobjjs1CDWUcQxndtnLkKAdrSrM",
        "3crBqZZsHhoLphM55MG4KRW6SbNzFEBFnehw7PW7ZRKt",
        "7fXKY9tPpvYsdbSNyesUqo27WYC6ZsBEULdtngGHqLCK",
        "J1S9H3QjnRtBbbuD4HjPV6RpRhwuk4zKbxsnCHuTgh9w",
        "8KyuwGzav7jTW9YaBGj2Qtp2q24zPUR3rD5caojXaby4",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;

    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    index_metadata_jsons(
        &setup,
        &[
            "AH6wj7T8Ke5nbukjtcobjjs1CDWUcQxndtnLkKAdrSrM",
            "J1S9H3QjnRtBbbuD4HjPV6RpRhwuk4zKbxsnCHuTgh9w",
        ],
        serde_json::from_str(SAMPLE_METADATA_JSON).unwrap(),
    )
    .await;

    let request = r#"
    {
        "groupKey": "collection",
        "groupValue": "J1S9H3QjnRtBbbuD4HjPV6RpRhwuk4zKbxsnCHuTgh9w",
        "displayOptions": {
            "showCollectionMetadata": true
        },
        "limit": 50,
        "page": 1
    }
    "#;

    let request: api::GetAssetsByGroup = serde_json::from_str(request).unwrap();

    let response = setup.das_api.get_assets_by_group(request).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_get_asset_by_owner_with_show_fungible() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "Ff3Ci7EnM6KUdNDtzZvCjC5APgz1ajYW2Pt2BfT6Wftq",
        "7mX5Gzny3r8WBRnnyQt6vVEkQ1AcN9twhPG1b5j5UhoC",
        "FQ3ePUyLt2UvuQCw2nBKG8LGLvXHAZPzpdD6QQAAv9pW",
        "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB",
        "4Av8W98PVbeYcdaw6T3L6STDdL5p3MWrqU7RkB8os4i7",
        "8c3zk1t1qt3RU43ckuvPkCS7HLbjJqq3J3Me8ov4aHrp",
        "J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn",
        "8yn5oqFMwYA8SgGqWwKq1Hia8aM5gh1DWmHEL34hMqBX",
        "4gbnMK8Yv5ySJ4aFFTR9EFf3nY2GKbDwd9U1zh41inX9",
        "HZ1JovNiVvGrGNiiYvEozEVgZ58xaU3RKwX8eACQBCt3",
        "Ah32VqAKNiE68qP35aPwxUbS15AuMo6d37ppyexR1CVe",
        "8LfiSAHzkKFsPCkwqYu11YtyhdN5dHTyDZc5trujPubg",
        "FoRGERiW7odcCBGU1bztZi16osPBHjxharvDathL5eds",
        "FahUFnmJDwVEcsRkrEmN2sZyEyjijQAwRfm7A4RCRou7",
        "5CUdD64Zktf2KJyHpaEKMzAVrNwG2saYX4FSsCheuLdA",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "ownerAddress": "GqPnSDXwp4JFtKS7YZ2HERgBbYLKpKVYy9TpVunzLRa9",
        "displayOptions": {
            "showFungible": true
        }
    }
    "#;

    let request: api::GetAssetsByOwner = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_assets_by_owner(request).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_get_asset_by_owner_without_show_fungible() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "Ff3Ci7EnM6KUdNDtzZvCjC5APgz1ajYW2Pt2BfT6Wftq",
        "7mX5Gzny3r8WBRnnyQt6vVEkQ1AcN9twhPG1b5j5UhoC",
        "FQ3ePUyLt2UvuQCw2nBKG8LGLvXHAZPzpdD6QQAAv9pW",
        "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB",
        "4Av8W98PVbeYcdaw6T3L6STDdL5p3MWrqU7RkB8os4i7",
        "8c3zk1t1qt3RU43ckuvPkCS7HLbjJqq3J3Me8ov4aHrp",
        "J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn",
        "8yn5oqFMwYA8SgGqWwKq1Hia8aM5gh1DWmHEL34hMqBX",
        "4gbnMK8Yv5ySJ4aFFTR9EFf3nY2GKbDwd9U1zh41inX9",
        "HZ1JovNiVvGrGNiiYvEozEVgZ58xaU3RKwX8eACQBCt3",
        "Ah32VqAKNiE68qP35aPwxUbS15AuMo6d37ppyexR1CVe",
        "8LfiSAHzkKFsPCkwqYu11YtyhdN5dHTyDZc5trujPubg",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "ownerAddress": "GqPnSDXwp4JFtKS7YZ2HERgBbYLKpKVYy9TpVunzLRa9"
    }
    "#;

    let request: api::GetAssetsByOwner = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_assets_by_owner(request).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_get_asset_with_show_inscription() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "9FkS3kZV4MoGps14tUSp7iVnizGbxcK4bDEhSoF5oYAZ",
        "HMixBLSkuhiGgVbcGhqJar476xzu1bC8wM7yHsc1iXwP",
        "DarH4z6SmdVzPrt8krAygpLodhdjvNAstP3taj2tysN2",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "id": "9FkS3kZV4MoGps14tUSp7iVnizGbxcK4bDEhSoF5oYAZ",
        "displayOptions": {
            "showInscription": true
        }
    }
    "#;

    let request: api::GetAsset = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_asset(request).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_show_zero_balance_filter_set_to_true() {
    let name = trim_test_name(function_name!());

    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "BE1CkzRjLTXAWcSVCaqzycwXsZ18Yuk3jMDMnPUoHjjS",
        "JUPyiwrYJFskUPiHa7hkeR8VUtAeFoSYbKedZNsDvCN",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "ownerAddress":"2oerfxddTpK5hWAmCMYB6fr9WvNrjEH54CHCWK8sAq7g",
        "displayOptions": {
            "showZeroBalance": true
        }
    }
    "#;

    let request: api::GetTokenAccounts = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_token_accounts(request).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_show_zero_balance_filter_being_set_to_false() {
    let name = trim_test_name(function_name!());

    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "BE1CkzRjLTXAWcSVCaqzycwXsZ18Yuk3jMDMnPUoHjjS",
        "JUPyiwrYJFskUPiHa7hkeR8VUtAeFoSYbKedZNsDvCN",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;

    let request = r#"        
    {
        "ownerAddress":"2oerfxddTpK5hWAmCMYB6fr9WvNrjEH54CHCWK8sAq7g",
        "displayOptions": {
            "showZeroBalance": false
        }
    }
    "#;

    let request: api::GetTokenAccounts = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_token_accounts(request).await.unwrap();

    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_show_unverified_collections_set_to_true() {
    let name = trim_test_name(function_name!());

    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "GAinsU7ajZa3m8M25rVT1hMgRWSh7NMSjUYGNC7UVRWU",
        "BMRVjv1deZ5WvSsk5sWQfbrh2LdRS8VK2nrsVbbEao4j",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect::<Vec<&SeedEvent>>()).await;

    let request = r#"        
    {
        "id":"GAinsU7ajZa3m8M25rVT1hMgRWSh7NMSjUYGNC7UVRWU",
        "displayOptions": {
            "showUnverifiedCollections": true
        }
    }
    "#;
    let request: api::GetAsset = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_asset(request).await.unwrap();
    insta::assert_json_snapshot!(name, response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_show_unverified_collections_set_to_false() {
    let name = trim_test_name(function_name!());

    let setup = TestSetup::new_with_options(
        name.clone(),
        TestSetupOptions {
            network: Some(Network::Mainnet),
        },
    )
    .await;

    let seeds: Vec<SeedEvent> = seed_accounts([
        "GAinsU7ajZa3m8M25rVT1hMgRWSh7NMSjUYGNC7UVRWU",
        "BMRVjv1deZ5WvSsk5sWQfbrh2LdRS8VK2nrsVbbEao4j",
    ]);

    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect::<Vec<&SeedEvent>>()).await;

    let request = r#"        
    {
        "id":"GAinsU7ajZa3m8M25rVT1hMgRWSh7NMSjUYGNC7UVRWU",
        "displayOptions": {
            "showUnverifiedCollections": false
        }
    }
    "#;
    let request: api::GetAsset = serde_json::from_str(request).unwrap();
    let response = setup.das_api.get_asset(request).await.unwrap();
    insta::assert_json_snapshot!(name, response);
}
