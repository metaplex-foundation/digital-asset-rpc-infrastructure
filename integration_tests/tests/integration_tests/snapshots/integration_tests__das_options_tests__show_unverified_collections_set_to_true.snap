---
source: integration_tests/tests/integration_tests/das_options_tests.rs
expression: response
---
{
  "interface": "V1_NFT",
  "id": "GAinsU7ajZa3m8M25rVT1hMgRWSh7NMSjUYGNC7UVRWU",
  "content": {
    "$schema": "https://schema.metaplex.com/nft1.0.json",
    "json_uri": "https://graphigo.prd.galaxy.eco/metadata/GZ9dKMp357jjMohEXrMcd2RJH9fGYq3boeJ2wttpQMuk/1965.json",
    "files": [],
    "metadata": {
      "name": "Olympian",
      "symbol": "olympian",
      "token_standard": "NonFungible"
    },
    "links": {}
  },
  "authorities": [
    {
      "address": "GZ9dKMp357jjMohEXrMcd2RJH9fGYq3boeJ2wttpQMuk",
      "scopes": [
        "full"
      ]
    }
  ],
  "compression": {
    "eligible": false,
    "compressed": false,
    "data_hash": "",
    "creator_hash": "",
    "asset_hash": "",
    "tree": "",
    "seq": 0,
    "leaf_id": 0
  },
  "grouping": [
    {
      "group_key": "collection",
      "group_value": "GZ9dKMp357jjMohEXrMcd2RJH9fGYq3boeJ2wttpQMuk",
      "verified": false
    }
  ],
  "royalty": {
    "royalty_model": "creators",
    "target": null,
    "percent": 0.0,
    "basis_points": 0,
    "primary_sale_happened": false,
    "locked": false
  },
  "creators": [
    {
      "address": "GZ9dKMp357jjMohEXrMcd2RJH9fGYq3boeJ2wttpQMuk",
      "share": 100,
      "verified": true
    }
  ],
  "ownership": {
    "frozen": false,
    "delegated": false,
    "delegate": null,
    "ownership_model": "single"
  },
  "supply": {
    "print_max_supply": 0,
    "print_current_supply": 0,
    "edition_nonce": 254
  },
  "mutable": true,
  "burnt": false,
  "token_info": {
    "supply": 1,
    "decimals": 0,
    "token_program": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
    "mint_authority": "GoEELrzpDpGYbqsbtUP2daadk8BSUhiPf3nzGaJFZHbR",
    "freeze_authority": "GoEELrzpDpGYbqsbtUP2daadk8BSUhiPf3nzGaJFZHbR"
  }
}
